### PR TITLE
add relational reader. add stepbuffer FFU. add test.

### PR DESCRIPTION
**Pull request by zalmane**

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -15,6 +15,16 @@
...
+    },
+    {
+      "name": "Python Debug: current",
+      "type": "python",
+      "request": "launch",
+      "env": {
+        "PYTHONDEVMODE": "1"
+      },
+      "program": "${workspaceFolder}/core/tests/tmp_test.py",
+      "justMyCode": false
```
Comment (spicy-sauce) [inline .vscode/launch.json:27]: we don't have such a test, consider removing this

File: core/src/datayoga_core/blocks/redis/write/block.py [hunk 2]
```diff
@@ -33,7 +33,8 @@ def init(self, context: Optional[Context] = None):
     async def run(self, data: List[Dict[str, Any]]) -> Tuple[List[Dict[str, Any]], List[Result]]:
         pipeline = self.redis_client.pipeline()
         for record in data:
-            dict_as_list = list(reduce(lambda x, y: x + y, record.items()))
+            # transform to a list, filtering it out None, which Redis does not support
+            dict_as_list = sum(filter(lambda i: i[1] is not None and i[0] != Block.MSG_ID_FIELD,record.items()),())
```
Comment (spicy-sauce) [inline core/src/datayoga_core/blocks/redis/write/block.py]: add spaces:
`dict_as_list = sum(filter(lambda i: i[1] is not None and i[0] != Block.MSG_ID_FIELD, record.items()), ())`

Comment (spicy-sauce) [inline core/src/datayoga_core/blocks/redis/write/block.py]: We probably should filter out by prefix as we also have the opcode internal field.

File: core/src/datayoga_core/blocks/relational/read/block.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
+import logging
+import os
```
Comment (spicy-sauce) [inline core/src/datayoga_core/blocks/relational/read/block.py]: not used

File: core/src/datayoga_core/blocks/relational/read/block.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+from csv import DictReader
```
Comment (spicy-sauce) [inline core/src/datayoga_core/blocks/relational/read/block.py]: not used

File: core/src/datayoga_core/blocks/relational/read/block.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+                host=connection.get("host"),
+                port=connection.get("port"),
+                username=connection.get("user"),
+                password=connection.get("password"),
+                database=connection.get("database")),
+            echo=connection.get("debug", False), connect_args=connection.get("connect_args", {}))
+        self.tbl = sa.Table(self.table, sa.MetaData(schema=self.schema), autoload_with=engine)
+
+        logger.debug(f"Connecting to {self.db_type}")
+        self.connection = engine.connect()
```
Comment (spicy-sauce) [inline core/src/datayoga_core/blocks/relational/read/block.py]: We probably should share this (relational_utils) as the `relational.write` does the same.

File: core/src/datayoga_core/blocks/relational/write/block.py [hunk 6]
```diff
@@ -1,10 +1,10 @@
...
-from enum import Enum, unique
 from typing import Any, Dict, List, Optional, Tuple
 
 import sqlalchemy as sa
 from datayoga_core import utils, write_utils
 from datayoga_core.block import Block as DyBlock
+from datayoga_core.blocks.relational import utils as relational_utils
 from datayoga_core.context import Context
 from datayoga_core.result import Result
 from sqlalchemy import Table
```
Comment (spicy-sauce) [inline core/src/datayoga_core/blocks/relational/write/block.py]: not in use

File: core/src/datayoga_core/job.py [hunk 7]
```diff
@@ -11,11 +10,13 @@
 from typing import Any, Dict, List, Optional, Tuple
 
 import jsonschema
+
```
Comment (spicy-sauce) [inline core/src/datayoga_core/job.py]: redundant, use isort, it will do it for you.

File: core/src/datayoga_core/job.py [hunk 8]
```diff
@@ -11,11 +10,13 @@
...
 from datayoga_core import blocks, utils
 from datayoga_core.block import Block
 from datayoga_core.context import Context
 from datayoga_core.result import Result, Status
 from datayoga_core.step import Step
+from datayoga_core.step_buffer import StepBuffer
```
Comment (spicy-sauce) [inline core/src/datayoga_core/job.py]: not in use

File: core/src/datayoga_core/step.py [hunk 9]
```diff
@@ -105,6 +105,7 @@ async def join(self):
     async def stop(self):
         # wait for all tasks to finish
         await self.join()
+        logger.debug("joined")
```
Comment (spicy-sauce) [inline core/src/datayoga_core/step.py:107]: maybe change to something more meaningful.

File: core/src/datayoga_core/step_buffer.py [hunk 10]
```diff
@@ -0,0 +1,55 @@
+import asyncio
+import datetime
```
Comment (spicy-sauce) [inline core/src/datayoga_core/step_buffer.py]: not in use

File: core/src/datayoga_core/step_buffer.py [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+import logging
+
+from datayoga_core.step import Step
+
+logger = logging.getLogger("dy")
+
+class StepBuffer(Step):
+    def __init__(self, id, min_buffer_size=4, max_buffer_size=4, flush_ms=None):
+        super().__init__(id, None)
```
Comment (spicy-sauce) [inline core/src/datayoga_core/step_buffer.py:10]: Change in Step so the Block will be optional:
`def __init__(self, id: str, block: Optional[Block], concurrency=1):`

File: core/src/datayoga_core/step_buffer.py [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+        self.min_buffer_size = min_buffer_size
+        self.max_buffer_size = max(max_buffer_size,min_buffer_size)
+        self.buffer = []
+        self.flush_ms = flush_ms
+        self.timer = None
+        self.concurrency_lock = asyncio.Semaphore(1)
+
+    async def flush_timer(self):
+        await asyncio.sleep(self.flush_ms/1000)
+        logger.debug(f"flushing on timeout")
```
Comment (spicy-sauce) [inline core/src/datayoga_core/step_buffer.py]: redundant f-string

File: core/src/datayoga_core/step_buffer.py [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+        super().__init__(id, None)
+        self.min_buffer_size = min_buffer_size
+        self.max_buffer_size = max(max_buffer_size,min_buffer_size)
+        self.buffer = []
+        self.flush_ms = flush_ms
+        self.timer = None
+        self.concurrency_lock = asyncio.Semaphore(1)
+
+    async def flush_timer(self):
+        await asyncio.sleep(self.flush_ms/1000)
```
Comment (spicy-sauce) [inline core/src/datayoga_core/step_buffer.py:19]: self.flush_ms can be None

File: core/src/datayoga_core/utils.py [hunk 14]
```diff
@@ -2,13 +2,16 @@
...
+import uuid
 from os import path
 from typing import Any, Dict, List, Tuple
 
 import yaml
+
 from datayoga_core import result
 from datayoga_core.block import Block
 from datayoga_core.context import Context
+from datayoga_core.producer import Message
```
Comment (spicy-sauce) [inline core/src/datayoga_core/utils.py]: not in use

File: core/src/datayoga_core/utils.py [hunk 15]
```diff
@@ -119,3 +122,6 @@ def reject_records(records: List[Dict[str, Any]], reason: str):
     for record in records:
         if not is_rejected(record):
             reject_record(reason, record)
+
+def add_uid(record: Dict[str, Any]) -> Dict[str,Any]:
+    return {Block.MSG_ID_FIELD: str(uuid.uuid4()), **record}
```
Comment (spicy-sauce) [inline core/src/datayoga_core/utils.py]: let's use f-string instead of str.

File: integration-tests/resources/jobs/tests/pg_to_redis.yaml [hunk 16]
```diff
@@ -0,0 +1,15 @@
+input:
+  uses: relational.read
+  with:
+    connection: psql-hr
+    schema: hr
+    table: emp
+steps:
+  # - uses: std.write
```
Comment (spicy-sauce) [inline integration-tests/resources/jobs/tests/pg_to_redis.yaml]: remove

File: integration-tests/test_pg_to_redis.py [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+    third_employee = redis_client.hgetall("12")
+    assert third_employee== {"id":"12",
+                            "full_name":"steve steve",
+                            "country": "972 - ISRAEL",
+                            "gender": "M",
+                            "address": "main street"}
+
+
+    redis_container.stop()
+
```
Comment (spicy-sauce) [inline integration-tests/test_pg_to_redis.py]: redundant empty line (IMO)

File: integration-tests/test_stdin_to_stdout copy.py [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+    run_job("tests.stdin_to_stdout", f"echo '{tested_data}\n{tested_data2}\n{tested_data3}'", output_file)
+
+    contents = open(output_file, "r").read()
+    result = [json.loads(str(item)) for item in contents.strip().split('\n')]
+
+    assert result == [
+        {"id": 121,"full_name": "Joe121 Allen", "country": "US - UNITED STATES", "gender": "M"},
+        {"id": 123,"full_name": "Joe123 Allen", "country": "US - UNITED STATES", "gender": "M"},
+    ]
+    os.remove(output_file)
```
Comment (spicy-sauce) [inline integration-tests/test_stdin_to_stdout copy.py]: remove this file

**Review by spicy-sauce — CHANGES_REQUESTED** [review]

File: core/src/datayoga_core/blocks/redis/write/block.py (hunk 2 above)
Comment (zalmane) [reply core/src/datayoga_core/blocks/redis/write/block.py]: done

File: core/src/datayoga_core/blocks/relational/read/block.py (hunk 3 above)
Comment (zalmane) [reply core/src/datayoga_core/blocks/relational/read/block.py]: done

File: core/src/datayoga_core/blocks/relational/read/block.py (hunk 5 above)
Comment (zalmane) [reply core/src/datayoga_core/blocks/relational/read/block.py]: done

File: core/src/datayoga_core/blocks/relational/write/block.py (hunk 6 above)
Comment (zalmane) [reply core/src/datayoga_core/blocks/relational/write/block.py]: done

File: core/src/datayoga_core/utils.py [hunk 19]
```diff
@@ -124,4 +122,4 @@ def reject_records(records: List[Dict[str, Any]], reason: str):
             reject_record(reason, record)
 
 def add_uid(record: Dict[str, Any]) -> Dict[str,Any]:
-    return {Block.MSG_ID_FIELD: str(uuid.uuid4()), **record}
+    return {Block.MSG_ID_FIELD: f"{uuid.uuid4()}", **record}
```
Comment (zalmane) [inline core/src/datayoga_core/utils.py:125]: are you sure it is faster?

File: core/src/datayoga_core/blocks/relational/read/block.py [hunk 20]
```diff
@@ -13,7 +13,7 @@
 
 class Block(DyProducer):
 
-    def init(self, context: Context):
+    def init(self, context: Optional[Context] = None):
```
Comment (zalmane) [inline core/src/datayoga_core/blocks/relational/read/block.py:16]: but here it is really not optional... I guess we can add a check and raise an exception if it is None

**Review by spicy-sauce — APPROVED** [review]

**Pull request merged**